### PR TITLE
refactor(app): set alert flag in state if U2E driver out of date

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR covers the state management aspects of #5491. It adds a very simple `alerts` module to the app's Redux state that:

- Can trigger an alert by placing an alert ID in state
- Can dismiss an alert by placing an alert ID in an `ignored` array in state
- Can dismiss an alert permanently by placing an alert ID in `alerts.ignored` in the config file

Additionally, it adds logic to the `system-info` module to:

- Trigger a `u2eDriverOutdated` alert if an out of date U2E driver is detected

In a future PR, a component will be added to the base `App` component to ensure that if there's an active alert in state, an alert modal for that given alert ID will be displayed, and the user will be able to dismiss the alert and/or tell the app "do not show this to me again".

## changelog

- refactor(app): set alert flag in state if U2E driver out of date

## review requests

This PR is difficult to test E2E without a Windows machine, but don't let that stop you from reviewing! The unit tests for this PR, given that the behavior is so computer setup specific, are quite important, so code reviews are very much appreciated.

If you're going to manually test this on a Windows machine, be sure to uninstall your U2E driver via Device Manager so that you will have an outdated driver the next time you plug the dongle in.

1. Launch the app on this branch
2. Plug the U2E adapter in if it's not plugged in already
    - If you are not on Windows or you don't have a dongle, you can simulate a dongle getting plugged in on Windows with an outdated driver by dispatching via the redux console:
    ```js
    {
        type: 'systemInfo:USB_DEVICE_ADDED',
        payload: {
            usbDevice: {
                locationId: 1,
                vendorId: 256,
                productId: 2560,
                deviceName: 'Realtek USB FE Family Controller',
                manufacturer: 'Realtek',
                serialNumber: 'Serial Number',
                deviceAddress: 5,
                windowsDriverVersion: '10.0.0.0',
            }
        }
    }
    ```
3. You should see:
    - [ ] `alerts:ALERT_TRIGGERED` action with `u2eDriverOutdated` in the payload
    - [ ] `state.alerts.active` should include `u2eDriverOutdated`
4. Trigger an alert dismissal through the redux console
    ```js
    {
        type: 'alerts:ALERT_DISMISSED',
        payload: { alertId: 'u2eDriverOutdated', remember: false },
    }
    ```
5. You should see
    - [ ] `state.alerts.active` should be empty
    - [ ] `state.alerts.ignored` should include `u2eDriverOutdated`
6. Trigger a permanent alert dismissal through the redux console
    ```js
    {
        type: 'alerts:ALERT_DISMISSED',
        payload: { alertId: 'u2eDriverOutdated', remember: true },
    }
    ```
7. Config updates should be triggered, such that:
    - [ ] `state.config.alerts.ignored` should include `u2eDriverOutdated`

## risk assessment

Low. App state additions are not currently wired to any UI, so even if things go wrong, they will not affect the UX. On top of that, all code additions are covered by unit tests.

### CODE REVIEW

File: app/src/alerts/epic.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { filter, withLatestFrom } from 'rxjs/operators'
+import union from 'lodash/union'
+
+import { getConfig, updateConfig } from '../config'
+import { ALERT_DISMISSED } from './constants'
+
+import type { Epic } from '../types'
+
+export const alertsEpic: Epic = (action$, state$) => {
+  return action$.pipe(
```
Comment (amitlissack) [inline app/src/alerts/epic.js:11]: Minor nit from a non-JS non-functional programmer:  Is alertsEpic a function that updates the ignored alerts config? Is there enough here (naming, comments, etc) to make it clear what this does?

File: app/src/system-info/epic.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import { map, mapTo, pairwise, filter } from 'rxjs/operators'
+
+import * as Alerts from '../alerts'
+import { OUTDATED } from './constants'
+import { getU2EWindowsDriverStatus } from './selectors'
+
+import type { Epic } from '../types'
+
+export const systemInfoEpic: Epic = (_, state$) => {
+  return state$.pipe(
```
Comment (amitlissack) [inline app/src/system-info/epic.js:11]: Minor nit from a non-JS non-functional programmer:  pretty much same comment as above.  what does this do? is it clear?

File: app/src/system-info/utils.js [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
+import { NOT_APPLICABLE, UNKNOWN, UP_TO_DATE, OUTDATED } from './constants'
+
+import type { UsbDevice, U2EAnalyticsProps, DriverStatus } from './types'
 
 const RE_REALTEK = /realtek/i
 
+// Driver version 10.38.117.2020, latest for Windows 10 as of 2020-04-12
+// NOTE(mc, 2020-05-05): this will cause false alerts on Windows 7; Realtek's
+// versioning scheme seems to be WindowsVersion.Something.Something.Year
+const REALTEK_UP_TO_DATE_VERSION = [10, 38, 117, 2020]
```
Comment (amitlissack) [inline app/src/system-info/utils.js:13]: Are we sure this should be in the code rather than a configuration file?

File: app/src/alerts/epic.js (hunk 1 above)
Comment (mcous) [reply app/src/alerts/epic.js:11]: Truthfully I'm not sure. As is unfortunately common with Redux, understanding what an individual function does requires an amount of ecosystem context. For Epics, you need to know (broadly) what an [RxJS Observable](https://rxjs.dev/guide/overview) is and what [redux-observable](https://redux-observable.js.org/docs/basics/Epics.html) does

- An Observable is, more or less, a type of push [Stream](https://en.wikipedia.org/wiki/Stream_(computing))
- redux-observable allows you to define epics which:
    - Are passed Observables of the the action stream (and corresponding state changes) of the app over time
        - `action$` and `stream$`, by convention where the trailing `$` says "this is an Obserable"
    - Return a new Observable that should emit actions that `redux-observable` will pass to the store

Once you've got that down, then you need to understand the RxJS's [pipe method and operators](https://rxjs.dev/guide/operators).

- `Observable.pipe` is a way to take one Observable and transform it in some way into a new `Observable`, as you might with an `Array.map` or `Array.filter`

So here, we:

1. Take our action observable
2. Transform it into a new observable which only emits actions that satisfy our filter condition that they are `ALERT_DISMISSED` an action and their `payload.remember` field is set
3. Transform it into a new observable that zips in the latest value from the `state$` stream and both the action value and the state value into an `updateConfig` action
4. We're left with an Observable that will, given the correct inputs, emit an `updateConfig` action to add an aler id to the config file's `alerts.ignored` array

If and when that `updateConfig` action is emitted from this Epic's observable, `redux-observable` will dispatch it, where other machinery will recognize the action and actually perform the side-effect of modifying the configuration file.

This definitely deserves some more comments because, if nothing else, the `withLatestFrom` is only in place because we don't have a previously defined action to append a value to an array in config, so we need to pull in state to get the existing config value to ensure we don't overwrite anything. It's... icky.

For the other stuff, @b-cooper could I maybe bug you for any strategies in your head for effective epic commenting + explanation?

File: app/src/system-info/epic.js (hunk 2 above)
Comment (mcous) [reply app/src/system-info/epic.js:11]: See above comment for background, but this one:

1. Takes the `state$` stream (so we can react to actual state changes so this epic doesn't need to care about which actions could trigger a given state change)
2. Map the state to a specific value computed from state with the `getU2EWindowsDriverStatus`
3. Transform the stream into a stream that collects emissions 2 at a time and groups them together
4. Filters our pairs of values to look for a change in value to the specific value `OUTDATED`
5. If that filter ever emits, we emit an `alertTriggered` action

Like in the example above, `redux-observable` will take that `alertTriggered` action and dispatch it into the store, where other (unit tested) machinery will do the actual state changes etc. Given that explanation, what comments might you want to see around this epic to make that operation more clear?

File: app/src/system-info/utils.js (hunk 3 above)
Comment (mcous) [reply app/src/system-info/utils.js:13]: Good point. At the very least I need a TODO here and a ticket. The short, not at all satisfying answer is that our app's config file doesn't currently have any sort of migration mechanism and I didn't want to deal with it hah

File: app/src/alerts/epic.js (hunk 1 above)
Comment (amitlissack) [reply app/src/alerts/epic.js:11]: Thanks for the explanation. Makes sense to me.

File: app/src/system-info/epic.js (hunk 2 above)
Comment (amitlissack) [reply app/src/system-info/epic.js:11]: I defer to your judgment, but I am a fan of comments. A one-liner that describes what the function does.

File: app/src/system-info/__tests__/selectors.test.js [hunk 4]
```diff
@@ -2,37 +2,68 @@
...
-    expected: null,
-  },
-  {
-    should: 'return a Realtek device with getU2EAdapterDevice',
-    selector: Selectors.getU2EAdapterDevice,
-    state: {
+describe('robot controls selectors', () => {
+  afterEach(() => {
+    jest.restoreAllMocks()
+  })
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/system-info/__tests__/selectors.test.js:13]: I don't believe so. This file only mocks with `jest.spyOn`, and after each test I want those spies completely removed. `mockReset` does not remove spies, `mockRestore` does (emphasis mine):

> mockFn.mockReset()
> 
> Does everything that mockFn.mockClear() does, and also removes any mocked return values or implementations.
>
> This is useful when you want to completely reset a mock back to its initial state. (Note that **resetting a spy will result in a function with no return value**).
>
> ...
> 
> mockFn.mockRestore()
>
> Does everything that mockFn.mockReset() does, and also restores the original (non-mocked) implementation.

File: app/src/alerts/__tests__/epic.test.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      expect(actual).toEqual(expected)
+    })
+  })
+
+  afterEach(() => {
+    jest.resetAllMocks()
+  })
+
+  it('should trigger a config:SET to save persistent alert ignores', () => {
+    getConfig.mockImplementation(state => {
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/alerts/__tests__/epic.test.js:34]: My opinions on mocking are _heavily_ influenced by the [testdouble.js library](https://github.com/testdouble/testdouble.js), especially its documentation on [stubbing](https://github.com/testdouble/testdouble.js/blob/master/docs/5-stubbing-results.md) and [verifying invocations](https://github.com/testdouble/testdouble.js/blob/master/docs/6-verifying-invocations.md)

`getConfig` is a pure function that provides data (as opposed to something that is generating a side-effect), so in "arrange-act-assert", I'd like to have my stub configuration be part of "arrange" rather than "assert". Unfortunately, jest doesn't have a nice stub configuration API where I can say "if provided this input, give me this output".

TL;DR: This `mockImplementation` with an `expect` is my crude way of setting up a stub in one go rather than mocking the result and then later asserting on how the mock was called.

With `testdouble`, this test would look like this:

```js
  it('should trigger a config:SET to save persistent alert ignores', () => {
    td.when(getConfig(MOCK_STATE)).thenReturn({
      alerts: { ignored: [MOCK_ALERT_1] }
    })

    testScheduler.run(({ hot, expectObservable }) => {
      const action$ = hot('-a', {
        a: Actions.alertDismissed(MOCK_ALERT_2, true),
      })
      const state$ = hot('s-', { s: MOCK_STATE })
      const output$ = alertsEpic(action$, state$)

      expectObservable(output$).toBe('-a', {
        a: Cfg.updateConfig('alerts.ignored', [MOCK_ALERT_1, MOCK_ALERT_2]),
      })
    })
  })
})

File: app/src/system-info/__tests__/utils.test.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
+// @flow
+
+import { mockUsbDevice, mockRealtekDevice } from '../__fixtures__'
+import { getDriverStatus } from '../utils'
+
+describe('system info utilities', () => {
+  describe('getDriverStatus', () => {
+    it('should return NOT_APPLICABLE if device is not Realtek', () => {
+      const device = mockUsbDevice
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/system-info/__tests__/utils.test.js:9]: The name of the test is indicative of the fact that there are three types of USB device in the eyes of our code:

1. Non-Realtek USB devices
2. Realtek USB devices
3. Realtek USB devices on Windows

It just so happens that one condition (`typeof windowsDriverVersion === 'undefined'`) is able to account for cases (1) and (2) simultaneously, but I don't think the current implementation details of the function make the test or the test name wrong. Happy to change it if you've got something you'd prefer, though!

[the remark this answers is not in the document]
Comment (mcous) [reply app/src/system-info/__tests__/utils.test.js:9]: Talked in Slack and added an extra check to the `getDriverStatus` implementation for robustness
